File: packages/server/src/collectors/activity-poller.ts
/**
 * PhantomOS Activity Poller
 * Tail-reads active sessions' JSONL files for real-time activity events.
 * @author Subash Karki
 */
import { logger } from '../logger.js';
import { openSync, readSync, closeSync, statSync } from 'node:fs';
import { basename, join } from 'node:path';
import { eq } from 'drizzle-orm';
import { db, sessions, activityLog } from '@phantom-os/db';
import { PROJECTS_DIR } from '@phantom-os/shared/constants-node';
import { safeReadDir } from '@phantom-os/shared/file-utils';

type Broadcast = (event: string, data: unknown) => void;

interface ActivityEvent {
  id: string;
  sessionId: string;
  sessionName: string;
  category: string;    // 'code', 'terminal', 'search', 'task', 'agent', 'git', 'user', 'response'
  icon: string;        // Lucide icon name
  message: string;
  detail?: string;     // Extra context (file path, command, etc.)
  timestamp: string;
}

const TOOL_EVENTS: Record<string, { icon: string; prefix: string; category: string }> = {
  'Read': { icon: 'file-text', prefix: 'Reading', category: 'code' },
  'Edit': { icon: 'pencil', prefix: 'Editing', category: 'code' },
  'Write': { icon: 'file-plus', prefix: 'Creating', category: 'code' },
  'MultiEdit': { icon: 'pencil', prefix: 'Multi-editing', category: 'code' },
  'Bash': { icon: 'terminal-square', prefix: 'Running', category: 'terminal' },
  'Grep': { icon: 'search', prefix: 'Searching', category: 'search' },
  'Glob': { icon: 'folder-search', prefix: 'Finding', category: 'search' },
  'Agent': { icon: 'bot', prefix: 'Spawning', category: 'agent' },
  'TaskCreate': { icon: 'list-plus', prefix: 'Task', category: 'task' },
  'TaskUpdate': { icon: 'check-square', prefix: 'Updated', category: 'task' },
  'WebSearch': { icon: 'globe', prefix: 'Searching web', category: 'search' },
  'WebFetch': { icon: 'download', prefix: 'Fetching', category: 'search' },
  'Skill': { icon: 'wand', prefix: 'Using skill', category: 'agent' },
};

// Track by file path, not sessionId (conversation ID changes on /clear)
const fileOffsets = new Map<string, number>();

/**
 * Find the most recently modified JSONL in the project directory matching a session's cwd.
 * After /clear, Claude creates a new JSONL with a different UUID, so we can't match by sessionId.
 * Instead, find the most recently modified .jsonl in the project dir.
 */
const findActiveJsonl = (cwd: string | null): string | null => {
  if (!cwd) return null;

  // Encode cwd to project dir name: /Users/subash.karki/CZ/foo → -Users-subash-karki-CZ-foo
  const encoded = cwd.replace(/\//g, '-');
  const projectDir = join(PROJECTS_DIR, encoded);

  try {
    const files = safeReadDir(projectDir)
      .filter((f) => f.endsWith('.jsonl') && !f.includes('/'))
      .map((f) => {
        const fullPath = join(projectDir, f);
        try {
          return { path: fullPath, mtime: statSync(fullPath).mtimeMs };
        } catch {
          return null;
        }
      })
      .filter((f): f is { path: string; mtime: number } => f !== null)
      .sort((a, b) => b.mtime - a.mtime);

    // Return most recently modified JSONL (that was touched in last 5 minutes)
    if (files.length > 0 && Date.now() - files[0].mtime < 5 * 60 * 1000) {
      return files[0].path;
    }
  } catch { /* project dir doesn't exist */ }
  return null;
};

const extractToolDetail = (toolName: string, input: Record<string, unknown>): string => {
  if (toolName === 'Read' || toolName === 'Edit' || toolName === 'Write') {
    const filePath = input.file_path as string | undefined;
    return filePath ? basename(filePath) : toolName;
  }
  if (toolName === 'Bash') {
    const command = input.command as string | undefined;
    return command ? command.slice(0, 60) : toolName;
  }
  if (toolName === 'Grep') {
    const pattern = input.pattern as string | undefined;
    return pattern ? pattern.slice(0, 60) : toolName;
  }
  if (toolName === 'Glob') {
    const pattern = input.pattern as string | undefined;
    return pattern ? pattern.slice(0, 60) : toolName;
  }
  if (toolName === 'Agent') {
    const description = input.description as string | undefined;
    return description ? description.slice(0, 60) : toolName;
  }
  if (toolName === 'TaskCreate') {
    const subject = input.subject as string | undefined;
    return subject ? subject.slice(0, 50) : toolName;
  }
  return toolName;
};

const extractEvents = (chunk: string, sessionId: string, sessionName: string): ActivityEvent[] => {
  const events: ActivityEvent[] = [];
  const lines = chunk.split('\n');

  for (const line of lines) {
    if (!line.trim()) continue;

    let entry: Record<string, unknown>;
    try {
      entry = JSON.parse(line);
    } catch {
      continue;
    }

    const type = entry.type as string | undefined;
    const timestamp = (entry.timestamp as string) ?? new Date().toISOString();

    if (type === 'assistant') {
      const message = entry.message as Record<string, unknown> | undefined;
      if (!message) continue;

      const content = message.content as unknown[] | undefined;
      if (!Array.isArray(content)) continue;

      for (const block of content) {
        if (!block || typeof block !== 'object') continue;
        const b = block as Record<string, unknown>;

        if (b.type === 'tool_use') {
          const toolName = b.name as string;
          const toolInput = (b.input as Record<string, unknown>) ?? {};
          const toolConfig = TOOL_EVENTS[toolName];

          // Detect git commits from Bash commands
          if (toolName === 'Bash') {
            const cmd = (toolInput.command as string) ?? '';
            const commitMatch = cmd.match(/git\s+commit\s+-m\s+["']?(.{1,60})/);
            if (commitMatch) {
              events.push({
                id: `act-${sessionId.slice(0, 6)}-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`,
                sessionId,
                sessionName,
                category: 'git',
                icon: 'git-commit-horizontal',
                message: `Committed: ${commitMatch[1].replace(/["']$/, '')}`,
                detail: cmd.slice(0, 120),
                timestamp,
              });
              continue;
            }
            // Detect git push
            if (/git\s+push/.test(cmd)) {
              events.push({
                id: `act-${sessionId.slice(0, 6)}-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`,
                sessionId,
                sessionName,
                category: 'git',
                icon: 'upload',
                message: 'Pushed to remote',
                detail: cmd.slice(0, 120),
                timestamp,
              });
              continue;
            }
            // Detect git checkout / branch
            const branchMatch = cmd.match(/git\s+(?:checkout|switch)\s+(?:-[bB]\s+)?(\S+)/);
            if (branchMatch) {
              events.push({
                id: `act-${sessionId.slice(0, 6)}-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`,
                sessionId,
                sessionName,
                category: 'git',
                icon: 'git-branch',
                message: `Branch: ${branchMatch[1]}`,
                detail: cmd.slice(0, 120),
                timestamp,
              });
              continue;
            }
          }

          // Resolve category/icon for known tools, MCP tools, and unknown tools
          let category: string;
          let icon: string;
          let displayName = toolName;
          let mcpServer: string | undefined;
          let mcpTool: string | undefined;

          if (toolConfig) {
            category = toolConfig.category;
            icon = toolConfig.icon;
          } else if (toolName.startsWith('mcp__')) {
            // MCP tool: mcp__<server>__<tool> → extract server + tool
            const parts = toolName.split('__');
            mcpServer = parts[1] ?? 'unknown';
            mcpTool = parts.slice(2).join('__') || 'unknown';
            displayName = `${mcpServer}:${mcpTool}`;
            category = 'mcp';
            icon = 'plug';
          } else {
            category = 'code';
            icon = 'circle';
          }

          const detail = extractToolDetail(toolName, toolInput);
          events.push({
            id: `act-${sessionId.slice(0, 6)}-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`,
            sessionId,
            sessionName,
            category,
            icon,
            message: toolConfig ? `${toolConfig.prefix} ${detail}` : `${displayName} ${detail}`,
            detail: toolName === 'Read' || toolName === 'Edit' || toolName === 'Write'
              ? (toolInput.file_path as string | undefined) ?? undefined
              : undefined,
            timestamp,
          });

          // Persist to DB for cross-session analytics
          try {
            db.insert(activityLog).values({
              timestamp: new Date(timestamp).getTime(),
              type: toolName,
              sessionId,
              metadata: JSON.stringify({
                category,
                detail,
                sessionName,
                displayName,
                skill: toolName === 'Skill' ? (toolInput.skill as string | undefined) : undefined,
                agentDesc: toolName === 'Agent' ? (toolInput.description as string | undefined) : undefined,
                mcpServer,
                mcpTool,
              }),
            }).run();
          } catch { /* non-critical — don't break the poller */ }
        } else if (b.type === 'text') {
          const text = b.text as string | undefined;
          if (!text) continue;
          if (text.startsWith('<')) continue;
          if (text.length < 15) continue;

          events.push({
            id: `act-${sessionId.slice(0, 6)}-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`,
            sessionId,
            sessionName,
            category: 'response',
            icon: 'message-square',
            message: text.slice(0, 80),
            timestamp,
          });
        }
      }
    }

    if (type === 'user') {
      const message = entry.message as Record<string, unknown> | undefined;
      if (!message) continue;

      const content = message.content;
      let text: string | null = null;

      if (typeof content === 'string') {
        text = content;
      } else if (Array.isArray(content) && content.length > 0) {
        const first = content[0] as Record<string, unknown> | undefined;
        if (first && typeof first.text === 'string') {
          text = first.text;
        }
      }

      if (!text) continue;
      if (text.startsWith('<command')) continue;
      if (text.startsWith('<system')) continue;

      events.push({
        id: `act-${sessionId.slice(0, 6)}-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`,
        sessionId,
        sessionName,
        category: 'user',
        icon: 'user',
        message: text.slice(0, 80),
        timestamp,
      });
    }
  }

  return events;
};

export const startActivityPoller = (broadcast: Broadcast): void => {
  const poll = () => {
    const activeSessions = db.select({ id: sessions.id, name: sessions.name, repo: sessions.repo, cwd: sessions.cwd })
      .from(sessions).where(eq(sessions.status, 'active')).all();

    // Accumulate events across ALL sessions, then emit a single broadcast per poll tick.
    // Without this, N active sessions can fire N separate SSE messages in one tick.
    const allEvents: ActivityEvent[] = [];

    for (const session of activeSessions) {
      // Find the actively-written JSONL by recent mtime, not by sessionId
      const jsonlPath = findActiveJsonl(session.cwd);
      if (!jsonlPath) continue;

      let fileSize: number;
      try { fileSize = statSync(jsonlPath).size; } catch { continue; }

      // Track by file path (different conversations have different files)
      const lastOffset = fileOffsets.get(jsonlPath) ?? Math.max(0, fileSize - 20_000);
      if (fileSize <= lastOffset) continue;

      const readSize = Math.min(fileSize - lastOffset, 100_000);
      const buffer = Buffer.alloc(readSize);
      const fd = openSync(jsonlPath, 'r');
      readSync(fd, buffer, 0, readSize, lastOffset);
      closeSync(fd);

      fileOffsets.set(jsonlPath, fileSize);

      const chunk = buffer.toString('utf-8');
      const sessionName = session.name ?? session.repo ?? session.id.slice(0, 8);
      const events = extractEvents(chunk, session.id, sessionName);

      allEvents.push(...events);
    }

    // Single broadcast per poll tick — cap at 100 events (across all sessions)
    if (allEvents.length > 0) {
      broadcast('activity', { events: allEvents.slice(-100) });
    }
  };

  setInterval(poll, 5_000);
  setTimeout(poll, 2_000);
  logger.info('ActivityPoller', 'Polling active sessions every 5s');
};

File: v2/internal/collector/activity_poller.go
// Package collector — ActivityPoller: real-time activity feed from JSONL files.
// Author: Subash Karki
package collector

import (
	"bufio"
	"context"
	"database/sql"
	"encoding/json"
	"fmt"
	"io"
	"log/slog"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"sync"
	"time"

	"github.com/fsnotify/fsnotify"
	"github.com/subashkarki/phantom-os-v2/internal/db"
	"github.com/subashkarki/phantom-os-v2/internal/provider"
)

const (
	activityPollInterval     = 5 * time.Second
	activityDebounce         = 500 * time.Millisecond
	activityFallbackInterval = 30 * time.Second
	activityRewatchInterval  = 60 * time.Second
)

// ActivityPoller watches active sessions' JSONL files for changes via fsnotify,
// extracts tool calls / git operations / messages, and inserts activity events.
// Falls back to 5s polling if fsnotify is unavailable.
type ActivityPoller struct {
	queries   *db.Queries
	prov      provider.Provider
	ctx       context.Context
	cancel    context.CancelFunc
	emitEvent func(name string, data interface{})
	journal   journalAppender // optional — appends notable events to journal work log

	mu          sync.Mutex
	fileOffsets map[string]int64 // byte offset per JSONL file path

	debounceMu     sync.Mutex
	debounceTimers map[string]*time.Timer
}

// activityEvent is a single parsed activity entry emitted per poll tick.
type activityEvent struct {
	SessionID string `json:"session_id"`
	Timestamp int64  `json:"timestamp"`
	Type      string `json:"type"`     // tool, git, message, response
	Icon      string `json:"icon"`     // emoji/icon hint for the frontend
	Category  string `json:"category"` // tool | git | message | response
	Detail    string `json:"detail"`   // file path, command, tool name, etc.
}

// NewActivityPoller creates a new ActivityPoller.
func NewActivityPoller(queries *db.Queries, prov provider.Provider, emitEvent func(string, interface{})) *ActivityPoller {
	return &ActivityPoller{
		queries:     queries,
		prov:        prov,
		emitEvent:   emitEvent,
		fileOffsets: make(map[string]int64),
	}
}

func (p *ActivityPoller) Name() string { return "activity-poller" }

// SetJournal injects the journal appender so the poller can log notable
// events (git commits, pushes, agent spawns) to the daily work log.
func (p *ActivityPoller) SetJournal(j journalAppender) {
	p.journal = j
}

func (p *ActivityPoller) Start(ctx context.Context) error {
	p.ctx, p.cancel = context.WithCancel(ctx)

	watcher, err := fsnotify.NewWatcher()
	if err != nil {
		slog.Warn("activity-poller: fsnotify unavailable, polling", "err", err)
		return p.pollLoop()
	}
	defer watcher.Close()

	p.debounceTimers = make(map[string]*time.Timer)
	p.watchProjectDirs(watcher)

	fallback := time.NewTicker(activityFallbackInterval)
	defer fallback.Stop()

	rewatch := time.NewTicker(activityRewatchInterval)
	defer rewatch.Stop()

	slog.Info("activity-poller started (fsnotify)")

	for {
		select {
		case <-p.ctx.Done():
			p.stopDebounceTimers()
			slog.Info("activity-poller stopped")
			return nil

		case ev, ok := <-watcher.Events:
			if !ok {
				return nil
			}
			if !strings.HasSuffix(ev.Name, ".jsonl") {
				continue
			}
			if !ev.Has(fsnotify.Write) && !ev.Has(fsnotify.Create) {
				continue
			}
			p.debouncedProcessFile(ev.Name)

		case err, ok := <-watcher.Errors:
			if !ok {
				return nil
			}
			slog.Error("activity-poller: watcher error", "err", err)

		case <-fallback.C:
			p.poll()

		case <-rewatch.C:
			p.watchProjectDirs(watcher)
		}
	}
}

// pollLoop is the legacy polling fallback when fsnotify is unavailable.
func (p *ActivityPoller) pollLoop() error {
	ticker := time.NewTicker(activityPollInterval)
	defer ticker.Stop()

	slog.Info("activity-poller started (polling)")

	for {
		select {
		case <-p.ctx.Done():
			slog.Info("activity-poller stopped")
			return nil
		case <-ticker.C:
			p.poll()
		}
	}
}

// watchProjectDirs adds fsnotify watches on conversations directory subdirectories.
func (p *ActivityPoller) watchProjectDirs(watcher *fsnotify.Watcher) {
	root := p.prov.ConversationsDir()
	entries, err := os.ReadDir(root)
	if err != nil {
		return
	}
	for _, e := range entries {
		if e.IsDir() {
			_ = watcher.Add(filepath.Join(root, e.Name()))
		}
	}
}

// debouncedProcessFile schedules processing for a JSONL file change after 500ms.
func (p *ActivityPoller) debouncedProcessFile(jsonlPath string) {
	p.debounceMu.Lock()
	defer p.debounceMu.Unlock()

	if t, ok := p.debounceTimers[jsonlPath]; ok {
		t.Stop()
	}
	p.debounceTimers[jsonlPath] = time.AfterFunc(activityDebounce, func() {
		p.processFileChange(jsonlPath)
		p.debounceMu.Lock()
		delete(p.debounceTimers, jsonlPath)
		p.debounceMu.Unlock()
	})
}

// processFileChange reads new lines from a single JSONL file and emits activity events.
func (p *ActivityPoller) processFileChange(jsonlPath string) {
	sessionID := sessionIDFromPath(jsonlPath)
	if sessionID == "" {
		return
	}
	events := p.readNewLines(jsonlPath, sessionID)
	if len(events) == 0 {
		return
	}
	p.persistAndEmit(events)
}

// stopDebounceTimers cancels all pending debounce timers during shutdown.
func (p *ActivityPoller) stopDebounceTimers() {
	p.debounceMu.Lock()
	defer p.debounceMu.Unlock()
	for k, t := range p.debounceTimers {
		t.Stop()
		delete(p.debounceTimers, k)
	}
}

func (p *ActivityPoller) Stop() error {
	if p.cancel != nil {
		p.cancel()
	}
	return nil
}

// poll iterates active sessions, finds their JSONL files, and processes new bytes.
func (p *ActivityPoller) poll() {
	sessions, err := p.queries.ListActiveSessions(p.ctx)
	if err != nil {
		slog.Error("activity-poller: list active sessions", "err", err)
		return
	}

	var allEvents []activityEvent

	// Track which JSONL paths are still active this tick for cleanup.
	activePaths := make(map[string]struct{}, len(sessions))

	for _, sess := range sessions {
		jsonlPath := p.findJSONLPath(sess)
		if jsonlPath == "" {
			continue
		}
		activePaths[jsonlPath] = struct{}{}

		events := p.readNewLines(jsonlPath, sess.ID)
		allEvents = append(allEvents, events...)
	}

	// Evict fileOffsets for paths that are no longer active so the map
	// doesn't grow without bound as sessions end.
	p.mu.Lock()
	for path := range p.fileOffsets {
		if _, ok := activePaths[path]; !ok {
			delete(p.fileOffsets, path)
		}
	}
	p.mu.Unlock()

	p.persistAndEmit(allEvents)
}

// persistAndEmit stores activity events in the DB and emits a Wails event.
func (p *ActivityPoller) persistAndEmit(events []activityEvent) {
	if len(events) == 0 {
		return
	}
	if len(events) > 100 {
		events = events[len(events)-100:]
	}

	now := time.Now().UnixMilli()
	for _, ev := range events {
		meta, _ := json.Marshal(map[string]string{
			"icon":     ev.Icon,
			"category": ev.Category,
			"detail":   ev.Detail,
		})
		_ = p.queries.InsertActivity(p.ctx, db.InsertActivityParams{
			Timestamp: now,
			Type:      ev.Type,
			SessionID: sql.NullString{String: ev.SessionID, Valid: true},
			Metadata:  sql.NullString{String: string(meta), Valid: true},
			Provider:  p.prov.Name(),
		})
	}

	if p.emitEvent != nil {
		p.emitEvent(EventActivity, events)
	}

	// Append notable events to the daily journal work log.
	// Only log git commits, pushes, and agent spawns — skip noisy events
	// like tool:read, tool:edit, message:assistant, etc.
	if p.journal != nil {
		today := time.Now().Format("2006-01-02")
		ts := time.Now().Format("15:04")
		for _, ev := range events {
			var line string
			switch ev.Type {
			case "git:commit":
				detail := ev.Detail
				if len(detail) > 120 {
					detail = detail[:120] + "..."
				}
				line = fmt.Sprintf("%s Committed: %s", ts, detail)
			case "git:push":
				line = fmt.Sprintf("%s Pushed to remote", ts)
			case "tool:agent":
				if ev.Detail != "" && ev.Detail != "Agent" {
					detail := ev.Detail
					if len(detail) > 100 {
						detail = detail[:100] + "..."
					}
					line = fmt.Sprintf("%s Spawned agent: %s", ts, detail)
				}
			}
			if line != "" {
				// Prefix with project name from session if available
				sess, err := p.queries.GetSession(p.ctx, ev.SessionID)
				if err == nil && sess.Repo.Valid && sess.Repo.String != "" {
					line = fmt.Sprintf("%s [%s] %s", ts, sess.Repo.String, line[6:]) // replace ts prefix
				}
				p.journal.AppendWorkLog(today, line)
			}
		}
	}
}

// findJSONLPath returns the conversation file path for a session using the
// provider's FindConversationFile method.
func (p *ActivityPoller) findJSONLPath(sess db.Session) string {
	cwd := sess.Cwd.String
	if cwd == "" {
		return ""
	}

	path, err := p.prov.FindConversationFile(sess.ID, cwd)
	if err != nil {
		slog.Debug("activity-poller: find conversation file", "session_id", sess.ID, "cwd", cwd, "err", err)
		return ""
	}
	return path
}

// readNewLines reads bytes from the last known offset, parses JSONL lines.
func (p *ActivityPoller) readNewLines(path, sessionID string) []activityEvent {
	p.mu.Lock()
	offset := p.fileOffsets[path]
	p.mu.Unlock()

	f, err := os.Open(path)
	if err != nil {
		slog.Debug("activity-poller: open jsonl", "path", path, "err", err)
		return nil
	}
	defer f.Close()

	// Seek to last known offset.
	if offset > 0 {
		if _, err := f.Seek(offset, io.SeekStart); err != nil {
			slog.Debug("activity-poller: seek", "path", path, "err", err)
			return nil
		}
	}

	var events []activityEvent
	scanner := bufio.NewScanner(f)
	// Allow up to 1 MB per line for large JSONL entries.
	scanner.Buffer(make([]byte, 0, 64*1024), 1024*1024)

	for scanner.Scan() {
		line := scanner.Bytes()
		if len(line) == 0 {
			continue
		}

		ev := p.parseLine(line, sessionID)
		if ev != nil {
			events = append(events, *ev)
		}
	}

	// Update offset to current position.
	newOffset, err := f.Seek(0, io.SeekCurrent)
	if err == nil {
		p.mu.Lock()
		p.fileOffsets[path] = newOffset
		p.mu.Unlock()
	}

	return events
}

// Known tool names to detect.
var knownTools = map[string]string{
	"Read":    "📖",
	"Edit":    "✏️",
	"Write":   "📝",
	"Bash":    "💻",
	"Grep":    "🔍",
	"Glob":    "📂",
	"Agent":   "🤖",
	"Skill":   "⚡",
	"WebFetch": "🌐",
}

// Git command patterns.
var gitCmdRegex = regexp.MustCompile(`git\s+(commit|push|checkout|branch|merge|rebase|pull|stash|reset)`)

// parseLine extracts an activity event from a single JSONL line.
func (p *ActivityPoller) parseLine(line []byte, sessionID string) *activityEvent {
	var raw map[string]interface{}
	if err := json.Unmarshal(line, &raw); err != nil {
		return nil
	}

	// Claude Code JSONL nests the actual payload under .message — unwrap so
	// helpers below can read content/role/input without a root-level wrapper.
	body := raw
	if msg, ok := raw["message"].(map[string]interface{}); ok {
		body = msg
	}

	// Use the outer envelope's type as the routing hint when present.
	outerType, _ := raw["type"].(string)

	now := time.Now().UnixMilli()

	// Check for tool use (assistant messages with tool_use content blocks).
	if isToolUse(body) {
		toolName := extractToolName(body)
		detail := extractToolDetail(body, toolName)
		icon, ok := knownTools[toolName]
		if !ok {
			icon = "🔧" // MCP or unknown tool
		}
		ev := &activityEvent{
			SessionID: sessionID,
			Timestamp: now,
			Type:      fmt.Sprintf("tool:%s", strings.ToLower(toolName)),
			Icon:      icon,
			Category:  "tool",
			Detail:    detail,
		}

		// If the tool is Bash and the command is a git op, emit a git event instead.
		if toolName == "Bash" {
			if cmd := extractBashContent(body); cmd != "" {
				if match := gitCmdRegex.FindStringSubmatch(cmd); len(match) > 1 {
					ev.Type = fmt.Sprintf("git:%s", match[1])
					ev.Icon = "🔀"
					ev.Category = "git"
					ev.Detail = truncate(cmd, 200)
				}
			}
		}
		return ev
	}

	// Check for user messages.
	if outerType == "user" || isUserMessage(body) {
		return &activityEvent{
			SessionID: sessionID,
			Timestamp: now,
			Type:      "message:user",
			Icon:      "💬",
			Category:  "message",
			Detail:    truncateStringField(body, "content", 120),
		}
	}

	// Check for assistant responses.
	if outerType == "assistant" || isAssistantMessage(body) {
		return &activityEvent{
			SessionID: sessionID,
			Timestamp: now,
			Type:      "message:assistant",
			Icon:      "🤖",
			Category:  "response",
			Detail:    truncateStringField(body, "content", 120),
		}
	}

	return nil
}

// --- Helpers for JSON introspection ---

func isToolUse(raw map[string]interface{}) bool {
	if t, _ := raw["type"].(string); t == "tool_use" {
		return true
	}
	if _, ok := raw["tool_name"]; ok {
		return true
	}
	// Nested content blocks.
	if content, ok := raw["content"].([]interface{}); ok {
		for _, block := range content {
			if m, ok := block.(map[string]interface{}); ok {
				if t, _ := m["type"].(string); t == "tool_use" {
					return true
				}
			}
		}
	}
	return false
}

func extractToolName(raw map[string]interface{}) string {
	if name, ok := raw["tool_name"].(string); ok {
		return name
	}
	if name, ok := raw["name"].(string); ok {
		return name
	}
	// Nested in content blocks.
	if content, ok := raw["content"].([]interface{}); ok {
		for _, block := range content {
			if m, ok := block.(map[string]interface{}); ok {
				if name, ok := m["name"].(string); ok {
					return name
				}
				if name, ok := m["tool_name"].(string); ok {
					return name
				}
			}
		}
	}
	return "unknown"
}

func extractToolDetail(raw map[string]interface{}, toolName string) string {
	// Try to extract file_path or command from the tool input.
	input, ok := raw["input"].(map[string]interface{})
	if !ok {
		// Try nested.
		if content, ok := raw["content"].([]interface{}); ok {
			for _, block := range content {
				if m, ok := block.(map[string]interface{}); ok {
					if inp, ok := m["input"].(map[string]interface{}); ok {
						input = inp
						break
					}
				}
			}
		}
	}
	if input == nil {
		return toolName
	}

	switch toolName {
	case "Read", "Edit", "Write":
		if fp, ok := input["file_path"].(string); ok {
			return fp
		}
	case "Bash":
		if cmd, ok := input["command"].(string); ok {
			return truncate(cmd, 200)
		}
	case "Grep", "Glob":
		if pat, ok := input["pattern"].(string); ok {
			return pat
		}
	}
	return toolName
}

func extractBashContent(raw map[string]interface{}) string {
	// Check if this is a Bash tool_use with a command input.
	toolName := extractToolName(raw)
	if toolName != "Bash" {
		return ""
	}
	input, ok := raw["input"].(map[string]interface{})
	if !ok {
		return ""
	}
	cmd, _ := input["command"].(string)
	return cmd
}

func isUserMessage(raw map[string]interface{}) bool {
	if t, _ := raw["type"].(string); t == "human" {
		return true
	}
	if role, _ := raw["role"].(string); role == "user" {
		return true
	}
	return false
}

func isAssistantMessage(raw map[string]interface{}) bool {
	if t, _ := raw["type"].(string); t == "assistant" {
		return true
	}
	if role, _ := raw["role"].(string); role == "assistant" {
		return true
	}
	return false
}

func truncateStringField(raw map[string]interface{}, key string, max int) string {
	s, _ := raw[key].(string)
	return truncate(s, max)
}

func truncate(s string, max int) string {
	if len(s) <= max {
		return s
	}
	return s[:max] + "..."
}
